File: tempestweb/native/cookies.py
"""Native cookies capability over ``document.cookie`` (Track N).

A typed Python awaitable mirror of the browser cookie jar, served by
``client/native/cookies.js`` in every mode (A via FFI, B via the ``native_call``
round-trip, C via the in-process facade). Only cookies visible to
``document.cookie`` are covered — ``HttpOnly`` cookies are, by design, invisible
to the client and must be set by the server response.

Conventions match the codebase rules: :func:`get` is a single-resource lookup that
returns ``None`` when the cookie is absent (not an error); :func:`all` is a
collection that returns ``{}`` when there are no cookies.
"""

from __future__ import annotations

from typing import Any

from tempestweb.native.dispatch import send_native_call

__all__ = ["all_cookies", "get", "remove", "set"]


async def get(name: str) -> str | None:
    """Read a single cookie by name.

    Args:
        name: The cookie name.

    Returns:
        The cookie's value, or ``None`` when no such cookie is readable.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    result = await send_native_call("cookies.get", {"name": name})
    value = result.get("value") if isinstance(result, dict) else result
    return None if value is None else str(value)


async def set(  # noqa: A001 — mirrors the browser's cookie "set" verb
    name: str,
    value: str,
    *,
    max_age: int | None = None,
    path: str = "/",
    same_site: str = "Lax",
    secure: bool = False,
) -> None:
    """Set a cookie.

    Args:
        name: The cookie name.
        value: The cookie value (percent-encoded on the wire).
        max_age: Lifetime in seconds; ``None`` makes it a session cookie.
        path: The cookie path scope (default ``"/"``).
        same_site: The ``SameSite`` policy (``"Lax"`` / ``"Strict"`` / ``"None"``).
        secure: Whether to mark the cookie ``Secure`` (HTTPS only).

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    await send_native_call(
        "cookies.set",
        {
            "name": name,
            "value": value,
            "max_age": max_age,
            "path": path,
            "same_site": same_site,
            "secure": secure,
        },
    )


async def remove(name: str, *, path: str = "/") -> None:
    """Remove a cookie by expiring it.

    Args:
        name: The cookie name.
        path: The path scope the cookie was set with (must match to clear it).

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    await send_native_call("cookies.remove", {"name": name, "path": path})


async def all_cookies() -> dict[str, str]:
    """Read every readable cookie.

    Returns:
        A name → value map (empty when there are no readable cookies).

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value: Any = await send_native_call("cookies.all", {})
    return {str(k): str(v) for k, v in dict(value or {}).items()}

File: client/native/cookies.js
// native/cookies.js — the browser side of the `cookies` capability.
//
// Reads and writes `document.cookie`. Mirrors tempestweb/native/cookies.py; the
// same handlers serve all three modes (A via FFI, B via the native_call round
// trip, C via the in-process facade). Non-HttpOnly cookies only — HttpOnly
// cookies are invisible to `document.cookie` by design.
//
// See ../../docs/contract.md and tempestweb/native/cookies.py.

import { CapabilityError } from "./index.js";

/**
 * Parse `document.cookie` into a name → value map (values percent-decoded).
 * @param {string} raw  The `document.cookie` string.
 * @returns {Object<string, string>}  The parsed cookie jar.
 */
function parseCookies(raw) {
  /** @type {Object<string, string>} */
  const jar = {};
  for (const part of (raw || "").split(";")) {
    const eq = part.indexOf("=");
    if (eq < 0) {
      continue;
    }
    const name = part.slice(0, eq).trim();
    if (name) {
      jar[name] = decodeURIComponent(part.slice(eq + 1).trim());
    }
  }
  return jar;
}

/**
 * Resolve the document that owns the cookie jar, or throw when unavailable.
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Document}
 */
function requireDocument(deps) {
  const doc = deps.document;
  if (!doc) {
    throw new CapabilityError("unsupported", "document.cookie is unavailable");
  }
  return doc;
}

/**
 * Read a single cookie by name.
 *
 * Returns a wrapper object (`{ value }`) rather than a bare string so the value
 * survives the dispatch layer's object normalization and an absent cookie
 * (`null`) is distinguishable from an empty one (`""`).
 *
 * @param {{name: string}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{value: ?string}>}  `{value}`, with `null` when absent.
 */
export async function cookiesGet(args, deps) {
  const jar = parseCookies(requireDocument(deps).cookie);
  const has = Object.prototype.hasOwnProperty.call(jar, args.name);
  return { value: has ? jar[args.name] : null };
}

/**
 * Read every readable cookie as a name → value map.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<Object<string, string>>}
 */
export async function cookiesAll(_args, deps) {
  return parseCookies(requireDocument(deps).cookie);
}

/**
 * Set a cookie.
 * @param {{name: string, value: string, max_age?: ?number, path?: string,
 *          same_site?: string, secure?: boolean}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<null>}
 */
export async function cookiesSet(args, deps) {
  const doc = requireDocument(deps);
  const parts = [`${args.name}=${encodeURIComponent(args.value ?? "")}`];
  parts.push(`Path=${args.path || "/"}`);
  if (args.max_age !== null && args.max_age !== undefined) {
    parts.push(`Max-Age=${args.max_age}`);
  }
  parts.push(`SameSite=${args.same_site || "Lax"}`);
  if (args.secure) {
    parts.push("Secure");
  }
  doc.cookie = parts.join("; ");
  return null;
}

/**
 * Remove a cookie by expiring it.
 * @param {{name: string, path?: string}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<null>}
 */
export async function cookiesRemove(args, deps) {
  const doc = requireDocument(deps);
  doc.cookie = `${args.name}=; Path=${args.path || "/"}; Max-Age=0`;
  return null;
}
